### PR TITLE
Implemented `GravitationalFreeSurfaceBC` for GPUs

### PR DESCRIPTION
**Pull request by ravil-mobile**
Tested `ocean` setup with CUDA/HIP/SYCL backends. Results obtained with CPUs and GPUs (with double precision) match up to `1e-9` after 100sec of the simulation 

Relevant PRs:
 - Device: https://github.com/SeisSol/Device/pull/19
 - yateto: https://github.com/SeisSol/yateto/pull/43

### CODE REVIEW

File: src/Initializer/BatchRecorders/DataTypes/Table.hpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  private:
+  std::vector<Type> hostVector{};
+  PointerType deviceDataPtr{nullptr};
+  device::DeviceInstance& device = device::DeviceInstance::getInstance();
+};
+
+/**
+ * This class may seem redundant. But it provides strong guarantee of
+ * zero initialization of std::array. Note, there are some circumstances
+ * when it is not zero-initialized
```
Comment (krenzland) [inline src/Initializer/BatchRecorders/DataTypes/Table.hpp:71]: This comment is a bit confusing. Do you mean to say that **without using this class** there might be some circumstances where std::array is not zero-init?

File: src/Initializer/BatchRecorders/LocalIntegrationRecorder.cpp [hunk 2]
```diff
@@ -171,8 +171,160 @@ void LocalIntegrationRecorder::recordDisplacements() {
...
+
+        (*currentTable)[key].set(inner_keys::Wp::Id::EasiBoundaryMap, easiBoundaryMapPtrs[face]);
+        (*currentTable)[key].set(inner_keys::Wp::Id::EasiBoundaryConstant,
+                                 easiBoundaryConstantPtrs[face]);
+      }
+    }
+  }
+}
+
+void LocalIntegrationRecorder::recordAnaliticalBc() {
```
Comment (krenzland) [inline src/Initializer/BatchRecorders/LocalIntegrationRecorder.cpp:303]: typo in function name, should be recordAnalyticalBc

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+  if (elementId < numElements) {
+    const double rho = rhos[elementId];
+    real* elementBoundaryDofs = dofsFaceBoundaryNodalPtrs[elementId];
+    real* elementDisplacement = displacementDataPtrs[elementId];
+
+    constexpr auto numNodes = seissol::nodal::tensor::nodes2D::Shape[0];
+    if (tid < numNodes) {
+      constexpr auto INodalStride = yateto::leadDim<seissol::init::INodal>();
+
+      const auto pressureAtBnd = real(-1.0) * rho * g * elementDisplacement[tid];
```
Comment (krenzland) [inline src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:29]: static_cast<real>(-1.0)

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+                                    real** easiBoundaryConstantPtrs,
+                                    size_t numElements) {
+
+  const int tid = threadIdx.x;
+  const int elementId = blockIdx.x;
+
+  constexpr auto leadINodalDim = yateto::leadDim<seissol::init::INodal>();
+  constexpr auto INodalDim0 = seissol::tensor::INodal::Shape[0];
+  constexpr auto INodalDim1 = seissol::tensor::INodal::Shape[1];
+  __shared__ __align__(8) real resultTerm[INodalDim1][INodalDim0];
```
Comment (krenzland) [inline src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:68]: does cuda support alignas(8)?

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+
+  constexpr auto leadConstantDim = yateto::leadDim<seissol::init::easiBoundaryConstant>();
+  constexpr auto ConstantDim0 = seissol::tensor::easiBoundaryConstant::Shape[0];
+  constexpr auto ConstantDim1 = seissol::tensor::easiBoundaryConstant::Shape[1];
```
Comment (krenzland) [inline src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:73]: should be constantDim0

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+  __shared__ __align__(8) real rightTerm[INodalDim1][leadConstantDim];
+
+  constexpr auto leadMapDim = yateto::leadDim<seissol::init::easiBoundaryMap>();
+  constexpr auto MapDim0 = seissol::tensor::easiBoundaryMap::Shape[0];
+  constexpr auto MapDim1 = seissol::tensor::easiBoundaryMap::Shape[1];
+  constexpr auto MapDim2 = seissol::tensor::easiBoundaryMap::Shape[2];
```
Comment (krenzland) [inline src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:79]: should be mapDim0

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+  auto stream = reinterpret_cast<cudaStream_t>(deviceStream);
+  kernelextractRotationMatrices<<<grid, block, 0, stream>>>(
+    displacementToFaceNormalPtrs,
+    displacementToGlobalDataPtrs,
+    TPtrs,
+    TinvPtrs,
+    numElements);
+}
+
+__global__  void kernelInitializeTaylorSeries(real** prevCoefficientsPtrs,
```
Comment (krenzland) [inline src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:189]: Maybe this should be named such that it is clear that it is the Taylor series for the grav. boundary?

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+  auto stream = reinterpret_cast<cudaStream_t>(deviceStream);
+  kernelInitializeTaylorSeries<<<grid, block, 0, stream>>>(
+    prevCoefficientsPtrs,
+    integratedDisplacementNodalPtrs,
+    rotatedFaceDisplacementPtrs,
+    deltaTInt,
+    numElements);
+}
+
+__global__ void kernelComputeInvImpedance(double* invImpedances,
```
Comment (krenzland) [inline src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:228]: kernelComputeInvAcousticImpedance would be better.

**Review by krenzland — CHANGES_REQUESTED** [review]
Looks great! I only found very few small things.
I've reviewed the cuda kernel first, comments also apply to hip/sycl kernels.

File: src/Equations/elastic/Kernels/DirichletBoundary.h [hunk 9]
```diff
@@ -65,6 +72,69 @@ class DirichletBoundary {
...
+    device.algorithms.incrementalAdd(
+      dofsFaceBoundaryNodalPtrs,
+      dofsFaceBoundaryNodalData,
+      tensor::INodal::size(),
+      numElements,
+      deviceStream
+    );
+
+    constexpr auto auxTmpMemSize = yateto::getMaxTmpMemRequired(nodalLfKrnlPrototype, projectKernelPrototype);
+    auto* auxTmpMem = (real*)(device.api->getStackMemory(auxTmpMemSize * numElements));
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/DirichletBoundary.h:102]: done

File: src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h [hunk 10]
```diff
@@ -164,8 +167,160 @@ class GravitationalFreeSurfaceBc {
...
+      kernels::time::aux::extractRotationMatrices(rotateDisplacementToFaceNormalPtrs,
+                                                  rotateDisplacementToGlobalPtrs,
+                                                  TDataPtrs,
+                                                  TinvDataPtrs,
+                                                  numElements,
+                                                  deviceStream);
+
+      auto rotateFaceDisplacementKrnl = kernel::gpu_rotateFaceDisplacement();
+      constexpr auto auxTmpMemSize = yateto::getMaxTmpMemRequired(rotateFaceDisplacementKrnl, projectKernelPrototype);
+      auto* auxTmpMem = (real*)(device.api->getStackMemory(auxTmpMemSize * numElements));
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h:228]: done

File: src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h [hunk 11]
```diff
@@ -164,8 +167,160 @@ class GravitationalFreeSurfaceBc {
...
+          std::make_tuple(&prevCoefficientsPtrs, &prevCoefficientsMemory, nodal::tensor::nodes2D::Shape[0])
+      };
+
+      const size_t numElements{dataTable[key].get(inner_keys::Wp::Id::Derivatives)->getSize()};
+      for (auto& item: dataPack) {
+        auto& ptrs = *std::get<0>(item);
+        auto& memory = *std::get<1>(item);
+        auto elementSize = std::get<2>(item);
+        memory = (real*)(device.api->getStackMemory(elementSize * numElements * sizeof(real)));
+        ptrs  = (real**)(device.api->getStackMemory(numElements * sizeof(real*)));
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h:212]: done

File: src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h [hunk 12]
```diff
@@ -164,8 +167,160 @@ class GravitationalFreeSurfaceBc {
...
+
+      auto** integratedDisplacementNodalPtrs = dataTable[key].get(inner_keys::Wp::Id::NodalAvgDisplacements)->getDeviceDataPtr();
+      kernels::time::aux::initializeTaylorSeries(prevCoefficientsPtrs,
+                                           integratedDisplacementNodalPtrs,
+                                           rotatedFaceDisplacementPtrs,
+                                           deltaTInt,
+                                           numElements,
+                                           deviceStream);
+
+      auto* invImpedances = (double*)(device.api->getStackMemory(sizeof(double) * numElements));
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h:251]: done

File: src/Equations/elastic/Kernels/Time.cpp [hunk 13]
```diff
@@ -286,15 +291,14 @@ void seissol::kernels::Time::computeBatchedAder(double i_timeStepWidth,
...
+                                        (entry.get(inner_keys::Wp::Id::Derivatives))->getDeviceDataPtr(),
                                         tensor::Q::Size,
                                         derivativesKrnl.numElements,
                                         device.api->getDefaultStream());
 
-    constexpr size_t MAX_TMP_MEM = (intKrnl.TmpMaxMemRequiredInBytes > derivativesKrnl.TmpMaxMemRequiredInBytes) \
-                                   ? intKrnl.TmpMaxMemRequiredInBytes : derivativesKrnl.TmpMaxMemRequiredInBytes;
-    real* tmpMem = (real*)(device.api->getStackMemory(MAX_TMP_MEM * NUM_ELEMENTS));
+    constexpr auto maxTmpMem = yateto::getMaxTmpMemRequired(intKrnl, derivativesKrnl);
+    real* tmpMem = (real*)(device.api->getStackMemory(maxTmpMem * numElements));
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/Time.cpp:301]: done

File: src/Initializer/BatchRecorders/DataTypes/Table.hpp (hunk 1 above)
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/DataTypes/Table.hpp:71]: I just removed the comment. The comment was outdated.

File: src/Initializer/BatchRecorders/LocalIntegrationRecorder.cpp (hunk 2 above)
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/LocalIntegrationRecorder.cpp:303]: done

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu (hunk 3 above)
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:29]: done

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu (hunk 6 above)
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:79]: done

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu (hunk 5 above)
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:73]: done

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu (hunk 7 above)
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu:189]: done

File: src/Initializer/BatchRecorders/DataTypes/EncodedConstants.hpp [hunk 14]
```diff
@@ -1,42 +1,64 @@
...
 #define SEISSOL_ENCODINGCONSTANTS_HPP
 
+#include "Kernels/precision.hpp"
 #include <cstdlib>
 
-namespace seissol {
-namespace initializers {
-namespace recording {
+namespace seissol::initializers::recording::inner_keys {
+struct Wp {
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Initializer/BatchRecorders/DataTypes/EncodedConstants.hpp:8]: Added struct comments in the doxygen style

File: src/Equations/elastic/Kernels/DirichletBoundary.h [hunk 15]
```diff
@@ -65,6 +72,69 @@ class DirichletBoundary {
...
+                        InverseMappingKrnl& nodalLfKrnlPrototype,
+                        local_flux::aux::DirichletBoundaryAux<Func>& boundaryCondition,
+                        ConditionalPointersToRealsTable &dataTable,
+                        device::DeviceInstance& device) const {
+
+    const size_t numElements{dataTable[key].get(inner_keys::Wp::Id::Dofs)->getSize()};
+
+    size_t memCounter{0};
+    auto* dofsFaceBoundaryNodalData = (real*)(device.api->getStackMemory(tensor::INodal::size() * numElements * sizeof(real)));
+    auto** dofsFaceBoundaryNodalPtrs = (real**)(device.api->getStackMemory(numElements * sizeof(real*)));
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/DirichletBoundary.h]: done

File: src/Equations/elastic/Kernels/Local.cpp [hunk 16]
```diff
@@ -94,9 +94,41 @@ void seissol::kernels::Local::setGlobalData(const CompoundGlobalData& global) {
...
   deviceLocalFluxKernelPrototype.fMrT = global.onDevice->localChangeOfBasisMatricesTransposed;
   deviceNodalLfKrnlPrototype.project2nFaceTo3m = global.onDevice->project2nFaceTo3m;
+
+  deviceProjectRotatedKrnlPrototype.V3mTo2nFace = global.onDevice->V3mTo2nFace;
 #endif
 }
 
+template<typename LocalDataType>
+struct ApplyAnalyticalSolution {
+  ApplyAnalyticalSolution(seissol::kernels::Local* kernel,
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/Local.cpp]: done

File: src/Equations/elastic/Kernels/LocalBase.h [hunk 17]
```diff
@@ -73,17 +73,22 @@ class seissol::kernels::LocalBase {
...
+    device::DeviceInstance& device = device::DeviceInstance::getInstance();
 #endif
 
     const std::vector<std::unique_ptr<physics::InitialField>> *initConds;
 public:
     virtual void setInitConds(decltype(initConds) initConds) {
       this->initConds = initConds;
     }
+
+    const std::unique_ptr<physics::InitialField>& getInitCond(size_t index) {
```
[the remark this answers is not in the document]
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/LocalBase.h]: done. Made the method to return a normal pointer to InitialField

**Comment by ravil-mobile** [comment]
Hi @krenzland, thanks for reviewing the code. Could you, please, have a look one more time?

File: src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu (hunk 4 above)
Comment (ravil-mobile) [reply src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu]: I don't know. In general, it is a piece of nonsense to align the shared memory. The compiler was supposed to do it. All in all,  `__align__(8)` was super crucial for A100 performance. I just don't want to even touch it.  Somehow `__align__(8)` significantly influences the instruction scheduling of `nvcc` compiler.

Comment (krenzland) [reply src/Equations/elastic/Kernels/DeviceAux/cuda/KernelsAux.cu]: Ah, I know this type of code...

**Review by krenzland — APPROVED** [review]

**Comment by ravil-mobile** [comment]
> Before merge, it'd be good if you could check if the submodules point to the correct commit. I think they do but it might be good to double check

I checked one more time and both `Device` and `yateto` (the ones which I edited) point to their corresponding master branches

**Comment by ravil-mobile** [comment]
@krenzland , can you review this yateto PR https://github.com/SeisSol/yateto/pull/44 ?

It turned out that there was an issue with `constexpr` with gcc@11

**Pull request merged**